### PR TITLE
wrap: Don't use --branch with shallow clones against HEAD

### LINKED ISSUE #10931: Git wrap errors incorrectly with shallow clones against HEAD
**Describe the bug**

The documentation states that Git wraps can be pointed to `revision = 'head'` to make them fetch the latest tip of the default branch.

https://github.com/mesonbuild/meson/blob/2761131ff94489842da3988792c4c2b0d25a850c/docs/markdown/Wrap-dependency-system-manual.md?plain=1#L112-L123

However, if one combines `revision = 'head'` with `depth = 1`, the following error occurs:

```
warning: Could not find remote branch HEAD to clone.
fatal: Remote branch HEAD not found in upstream origin

..\meson.build:1864:4: ERROR: Git command failed: ['C:\\Program Files\\Git\\cmd\\git.EXE', '-c', 'advice.detachedHead=false', 'clone', '--depth', '1', '--branch', 'HEAD', 'https://github.com/FFmpeg/gas-preprocessor.git', 'win-gas']
```

This bug was introduced by https://github.com/mesonbuild/meson/commit/507cf479411050b76dc8724de1b3b464b8f965e3, where having `depth = 1`  hardcodes the usage of `--branch`, ignoring if `revision` is also `head`:

https://github.com/mesonbuild/meson/blob/9a57608cac4a422ffa450bcf75c0609639035e11/mesonbuild/wrap/wrap.py#L578-L586

**To Reproduce**

meson.build:
```meson
  project('test', 'c')
  gas = find_program('gas-preprocessor.pl', required: false)

  if not gas.found()
    subproject('win-gas')
  endif
```

subprojects/win-gas.wrap:
```ini
[wrap-git]
url = https://github.com/FFmpeg/gas-preprocessor.git
revision = head
depth = 1

[provide]
program_names = gas-preprocessor.pl
```

**Expected behavior**
Git clone must succeed, and the resulting clone must be shallow and point to the tip of the `master` branch.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? **No**
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) **Windows 10 21H2**
* what Python version are you using e.g. 3.8.0 **3.10.7**
* what `meson --version` **0.63.2**
* what `ninja --version` if it's a Ninja build **1.10.2**

### PR DESCRIPTION
**Pull request by amyspark**
:wave:

This PR fixes #10931 by not adding `--branch` in the mentioned circumstances. I've added a new test to ensure it gets caught in the future.

### CODE REVIEW

File: mesonbuild/wrap/wrap.py [hunk 1]
```diff
@@ -581,9 +581,12 @@ def get_git(self) -> None:
                     if not verbose_git(checkout_cmd, self.dirname):
                         verbose_git(['fetch', self.wrap.get('url'), revno], self.dirname, check=True)
                         verbose_git(checkout_cmd, self.dirname, check=True)
-            else:
-                verbose_git(['-c', 'advice.detachedHead=false', 'clone', *depth_option, '--branch', revno, self.wrap.get('url'),
-                             self.directory], self.subdir_root, check=True)
+            elif revno.lower != 'head':
+                args = ['-c', 'advice.detachedHead=false', 'clone', *depth_option]
+                if revno.lower != 'head':
```
[the remark this answers is not in the document]
Comment (amyspark) [reply mesonbuild/wrap/wrap.py:586]: Rebase conflict got me there... the check's meant to add the `--branch` if `revno.lower` is not `HEAD`.

[the remark this answers is not in the document]
Comment (xclaesse) [reply mesonbuild/wrap/wrap.py:586]: Ok, makes much more sense now indeed.

**Comment by xclaesse** [comment]
CI fails, but other than that LGTM

**Comment by amyspark** [comment]
https://github.com/mesonbuild/meson/actions/runs/3632042330/jobs/6127422477#step:5:13

I don't understand this one, but the other linter does give an actionable message that I've fixed.

**Comment by xclaesse** [comment]
> I don't understand this one, but the other linter does give an actionable message that I've fixed.

Seems you figured it out, linter passed now.

**Pull request merged**